### PR TITLE
Validate SRUN variables when launching in SLURM

### LINKED ISSUE #10150: Validate environment variables in SLURMEnvironment and warn user about incompatible settings
## 🚀 Feature

Check slurm environment settings and print warnings if needed.

### Motivation

If SLURM srun variables are set incorrectly, the processes can hang and the user will not know why.

Examples:
- https://github.com/PyTorchLightning/pytorch-lightning/issues/4612#issuecomment-949937617
- #15141

### Pitch

In the SLURM cluster environment, check for example:

```py
if os.environ.get("SLURM_CPUS_PER_TASK") > 1:
    warn(
        "You asked SLURM for multiple CPUs but we are not sure your machine has multiple CPUs. "
        "If your process hangs, consider setting --cpus-per-task=1"
    )

```
### Alternatives

Do nothing. Users will keep submitting issues :) 

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

cc @borda @awaelchli

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Addresses suggestion in #14986, frequent user errors that result in their SLURM jobs hanging. By adding warnings and errors, we can prevent these frequent questions and give feedback directly at runtime.

Fixes #10150

No intended breaking changes:
- Lightning SLURM users were always meant to use the --ntasks-per-node, otherwise their processes would hang. Hence, the error is appropriate.
- The warning about srun not being used is very unlikely going to be a false positive, because the `srun` program is only installed on SLURM clusters. However, we won't make it an error just in case there is another program called `srun` on your system unrelated to SLURM.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @awaelchli @carmocca @justusschock

### CODE REVIEW

File: src/lightning_lite/plugins/environments/slurm_environment.py [hunk 1]
```diff
@@ -141,3 +144,19 @@ def resolve_root_node_address(nodes: str) -> str:
...
+        Right now, we only check for the most commond user errors. See `the srun docs
+        <https://slurm.schedmd.com/srun.html>`_ for a complete list of supported srun variables.
+        """
+
+        ntasks = int(os.environ.get("SLURM_NTASKS", "1"))
+        if ntasks > 1:
+            raise rank_zero_warn(
+                f"You set `--ntasks={ntasks}` in your SLURM bash script, but this variable is not supported. Please use"
+                f" `--ntasks-per-node={ntasks}` instead."
+            )
```
Comment (awaelchli) [inline src/lightning_lite/plugins/environments/slurm_environment.py:162]: @Queuecumber Sorry for pinging you, but as a slurm user you might know this. 
1. Should this be an error? For now I made it a warning as I am not 100% sure if I break someone's code.
2. Would it be useful to also provide a warning/error when SLURM_CPUS_PER_TASK is set, with an error similar as in #10150?

Comment (Queuecumber) [reply src/lightning_lite/plugins/environments/slurm_environment.py:162]: I won't claim any special knowledge of slurm outside whatever I need to run my experiments, but my understanding is that `ntasks` and `ntasks-per-node` can both be set and then `ntasks-per-node` functions as a maximum. Why you would want to do that and frankly what that even means I don't know but technically it's "allowed". So in that case I suppose a loud warning is best so that the one person who uses both those options and lightning doesn't have their code crash. 

That said, it may not be the best idea to commit to slurm validation. 

I always viewed the slurm multiprocessing detection/checkpointing/requeueing support in lightning as nice-haves with a limited scope and I think that fits nicely with the rest of what lightning does in terms of making your life easier without being too invasive. For example, lightning isn't writing the sbatch file and submitting the job for you; that would be out of scope.

It may be worth discussing if this kind of validation, too, is out of scope. I won't say that I think it definitely is but it might be worth discussing it internally. Validating all the permutations of slurm options and which ones conflict and how they change based on context is extremely complex and there are other options besides these that people may want validated. The other way to do it is to just go all-in and implement something like submitit which creates the sbatch file based on the trainer config, and that way you have complete control of what options are being used so you wouldn't need any validation logic.

Just something to think about

Comment (awaelchli) [reply src/lightning_lite/plugins/environments/slurm_environment.py:162]: Greatly appreciate the quick response!

>  but my understanding is that ntasks and ntasks-per-node can both be set and then ntasks-per-node functions as a maximum

I would say yes, that's also my understanding that this is allowed in general. But I don't see that being useful or compatible with Lightning Trainer at the moment. Right now, there is no elasticity and we require each node to run the same number of processes. I don't see that changing anytime soon. Since that number set there must also be exactly the same as in the Trainer(devices=x), I don't see how setting both `ntasks-per-node` and `ntasks` to different values is valid. An example of this mistake can be seen here: #14986, and similar questions have been asked in the past and on slack. 

> Validating all the permutations of slurm options and which ones conflict and how they change based on context is extremely complex and there are other options besides these that people may want validated. 

Yeah, I don't think that's feasible. That's why I wrote in the comment in the code that we only intend to catch the most frequent user mistakes. 

> The other way to do it is to just go all-in and implement something like submitit which creates the sbatch file based on the trainer config, and that way you have complete control of what options are being used so you wouldn't need any validation logic.

Out of scope of Lightning I think. I would rather invest in a smooth interaction with submitit and Lightning.

Comment (Queuecumber) [reply src/lightning_lite/plugins/environments/slurm_environment.py:162]: In that case I think crashing the job does make sense

File: src/lightning_lite/plugins/environments/slurm_environment.py [hunk 2]
```diff
@@ -89,6 +96,7 @@ def detect() -> bool:
         This will then avoid the detection of ``SLURMEnvironment`` and another environment can be detected
         automatically.
         """
+        SLURMEnvironment._validate_srun_used()
```
Comment (justusschock) [inline src/lightning_lite/plugins/environments/slurm_environment.py:99]: isn't it sufficient to only run this on the environment's init? Why would you want to run it also on detection? If this is used in the end, it will run in init anyways and otherwise it doesn't make sense to validate IMO

File: src/lightning_lite/plugins/environments/slurm_environment.py [hunk 3]
```diff
@@ -141,3 +149,36 @@ def resolve_root_node_address(nodes: str) -> str:
...
+    @staticmethod
+    def _validate_srun_used() -> None:
+        """Checks if the `srun` command is available and used.
+
+        Parallel jobs (multi-GPU, multi-node) in SLURM are launched by prepending `srun` in front of the Python command.
+        Not doing so will result in processes hanging, which is a frequent user error. Lightning will emit a warning if
+        `srun` is found but not used.
+        """
+        srun_exists = subprocess.call(["which", "srun"]) == 0
+        srun_used = "SLURM_NTASKS" in os.environ
```
Comment (justusschock) [inline src/lightning_lite/plugins/environments/slurm_environment.py:162]:
```suggestion
        srun_used = SLURMEnvironment.detect()
```

when not calling in `detect` (see my comment above), we could reuse this in case the implementation might change :)

File: src/lightning_lite/plugins/environments/slurm_environment.py [hunk 4]
```diff
@@ -141,3 +149,36 @@ def resolve_root_node_address(nodes: str) -> str:
...
+        hint = " ".join(["srun", os.path.basename(sys.executable), *sys.argv])[:64]
```
Comment (justusschock) [inline src/lightning_lite/plugins/environments/slurm_environment.py:163]: why do we limit this to 64 chars?

**Review by justusschock — APPROVED** [review]

File: src/lightning_lite/plugins/environments/slurm_environment.py (hunk 2 above)
Comment (awaelchli) [reply src/lightning_lite/plugins/environments/slurm_environment.py:99]: No, I think not. That would render the check useless. The SLURM environment variables are only properly set if srun is used. The dependencies are:
srun used -> environment variables set -> detect() returns True

So, when in the AcceleratorConnector we do 

```
for env in environments:
    env.detect()
```

we want that detect() checks if srun is available but not used. If it is available and not used, a different environment gets selected because SLURMEnvironment.detect() = False and the process hangs, but the warning gets emitted and the user can see it. 

On the other hand, we also call the check in init in case user explicitly instantiates and passes the plugin.

File: src/lightning_lite/plugins/environments/slurm_environment.py (hunk 3 above)
Comment (awaelchli) [reply src/lightning_lite/plugins/environments/slurm_environment.py:162]: This would result in infinite recursion, since we need to check it in detect() itself too.

File: src/lightning_lite/plugins/environments/slurm_environment.py (hunk 4 above)
Comment (awaelchli) [reply src/lightning_lite/plugins/environments/slurm_environment.py:163]: In case the python command is long (maybe many hyperparameters) we truncate it so the warning message remains readable. 64 is an arbitrary choice and the author likes powers of 2.

File: src/lightning_lite/plugins/environments/slurm_environment.py (hunk 3 above)
Comment (carmocca) [reply src/lightning_lite/plugins/environments/slurm_environment.py:162]: srun_used could become an utility fn in this file to reuse the implementation, where  both detect and this one call it.

File: src/lightning_lite/plugins/environments/slurm_environment.py [hunk 5]
```diff
@@ -141,3 +149,36 @@ def resolve_root_node_address(nodes: str) -> str:
         nodes = re.sub(r"\[(.*?)[,-].*\]", "\\1", nodes)  # Take the first node of every node range
         nodes = re.sub(r"\[(.*?)\]", "\\1", nodes)  # handle special case where node range is single number
         return nodes.split(" ")[0].split(",")[0]
+
+    @staticmethod
```
Comment (carmocca) [inline src/lightning_lite/plugins/environments/slurm_environment.py:153]: I suggest making this a utility function in this file. Unless you have a specific reason to make it static.

File: src/lightning_lite/plugins/environments/slurm_environment.py (hunk 4 above)
Comment (carmocca) [inline src/lightning_lite/plugins/environments/slurm_environment.py:163]: The hint should only be constructed if the `if` below is True

File: src/lightning_lite/plugins/environments/slurm_environment.py [hunk 6]
```diff
@@ -141,3 +149,36 @@ def resolve_root_node_address(nodes: str) -> str:
...
+        """Checks if the `srun` command is available and used.
+
+        Parallel jobs (multi-GPU, multi-node) in SLURM are launched by prepending `srun` in front of the Python command.
+        Not doing so will result in processes hanging, which is a frequent user error. Lightning will emit a warning if
+        `srun` is found but not used.
+        """
+        srun_exists = subprocess.call(["which", "srun"]) == 0
+        srun_used = "SLURM_NTASKS" in os.environ
+        hint = " ".join(["srun", os.path.basename(sys.executable), *sys.argv])[:64]
+        if srun_exists and not srun_used:
```
Comment (carmocca) [inline src/lightning_lite/plugins/environments/slurm_environment.py:164]: Should we also check if `srun` is in sys.argv? Or too brittle?

File: src/lightning_lite/plugins/environments/slurm_environment.py [hunk 7]
```diff
@@ -141,3 +149,36 @@ def resolve_root_node_address(nodes: str) -> str:
...
+            rank_zero_warn(
+                "The `srun` command is available on your system but is not used. HINT: If your intention is to run"
+                f" Lightning on SLURM, prepend your python command with `srun` like so: {hint} ...",
+                category=PossibleUserWarning,
+            )
+
+    @staticmethod
```
Comment (carmocca) [inline src/lightning_lite/plugins/environments/slurm_environment.py:171]: Same suggestion

File: src/lightning_lite/plugins/environments/slurm_environment.py [hunk 8]
```diff
@@ -141,3 +149,36 @@ def resolve_root_node_address(nodes: str) -> str:
...
+    def _validate_srun_variables() -> None:
+        """Checks for conflicting or incorrectly set variables set through `srun` and raises a useful error
+        message.
+
+        Right now, we only check for the most common user errors. See `the srun docs
+        <https://slurm.schedmd.com/srun.html>`_ for a complete list of supported srun variables.
+        """
+        ntasks = int(os.environ.get("SLURM_NTASKS", "1"))
+        if ntasks > 1 and "SLURM_NTASKS_PER_NODE" not in os.environ:
```
Comment (carmocca) [inline src/lightning_lite/plugins/environments/slurm_environment.py:180]: Why is ntasks okay if `SLURM_NTASKS_PER_NODE` is also set?

The error isn't clear, it says it's not supported but technically it is if both are set? The tests modified below have both set

**Review by carmocca — APPROVED** [review]

File: src/lightning_lite/plugins/environments/slurm_environment.py (hunk 5 above)
Comment (awaelchli) [reply src/lightning_lite/plugins/environments/slurm_environment.py:153]: The reason is that this is not really a utility function that should be used anywhere else. It only makes sense in the context of this SLURM class and therefore is a staticmethod.

File: src/lightning_lite/plugins/environments/slurm_environment.py (hunk 6 above)
Comment (awaelchli) [reply src/lightning_lite/plugins/environments/slurm_environment.py:164]: srun won't be in argv as argv only stores what is passed to python afaik. Anything before the python command won't show up.

File: src/lightning_lite/plugins/environments/slurm_environment.py (hunk 5 above)
Comment (carmocca) [reply src/lightning_lite/plugins/environments/slurm_environment.py:153]: This is mostly a stylistic choice, but so far we've been choosing module (aka utility) functions over static methods unless inheritance or a shared interface is desired. Not a strong suggestion though

File: src/lightning_lite/plugins/environments/slurm_environment.py (hunk 8 above)
Comment (awaelchli) [reply src/lightning_lite/plugins/environments/slurm_environment.py:180]: NTASKS is equivalent to WORLD_SIZE and should be there always. The key here is that the user instructs SLURM that it should start a fixed number of processes per node, and not distribute the processes arbitrarily. It is important that Lighting has the same number of processes on each node (the same limitation applies to torchrun/torchelastic). This is what we are requiring here.

In the future this might change (#14078) but currently it is a requirement.

File: src/lightning_lite/plugins/environments/slurm_environment.py (hunk 3 above)
Comment (awaelchli) [reply src/lightning_lite/plugins/environments/slurm_environment.py]: I did this, although I find it adds an indirection that makes the code less readable.
